File: bindings/java/src/main/java/com/dyber/pqc/KEM.java
/*
 * libpqc-dyber - Post-Quantum Cryptography Library
 * Copyright (c) 2024-2026 Dyber, Inc.
 * SPDX-License-Identifier: Apache-2.0 OR MIT
 */

package com.dyber.pqc;

/**
 * Key Encapsulation Mechanism (KEM) wrapper.
 *
 * <p>Provides key generation, encapsulation, and decapsulation operations
 * for post-quantum KEM algorithms.</p>
 *
 * <pre>{@code
 * try (KEM kem = new KEM("ML-KEM-768")) {
 *     KEM.KeyPair kp = kem.keygen();
 *     KEM.EncapsResult enc = kem.encaps(kp.publicKey);
 *     byte[] ss = kem.decaps(enc.ciphertext, kp.secretKey);
 *     // enc.sharedSecret equals ss
 * }
 * }</pre>
 */
public class KEM implements AutoCloseable {

    private long handle;

    /**
     * Result of a key generation operation.
     */
    public static class KeyPair {
        /** The public key bytes. */
        public final byte[] publicKey;
        /** The secret key bytes. */
        public final byte[] secretKey;

        KeyPair(byte[] publicKey, byte[] secretKey) {
            this.publicKey = publicKey;
            this.secretKey = secretKey;
        }
    }

    /**
     * Result of an encapsulation operation.
     */
    public static class EncapsResult {
        /** The ciphertext bytes. */
        public final byte[] ciphertext;
        /** The shared secret bytes. */
        public final byte[] sharedSecret;

        EncapsResult(byte[] ciphertext, byte[] sharedSecret) {
            this.ciphertext = ciphertext;
            this.sharedSecret = sharedSecret;
        }
    }

    /**
     * Create a new KEM context for the specified algorithm.
     *
     * @param algorithm the algorithm name (e.g., "ML-KEM-768")
     * @throws PQCException if the algorithm is not supported
     */
    public KEM(String algorithm) throws PQCException {
        this.handle = PQC.nativeKemNew(algorithm);
        if (this.handle == 0) {
            throw new PQCException(PQCException.NOT_SUPPORTED,
                "Unsupported KEM algorithm: " + algorithm);
        }
    }

    private void ensureOpen() {
        if (handle == 0) {
            throw new IllegalStateException("KEM context has been closed");
        }
    }

    /**
     * Return the algorithm name.
     */
    public String algorithm() {
        ensureOpen();
        return PQC.nativeKemAlgorithm(handle);
    }

    /**
     * Return the public key size in bytes.
     */
    public int publicKeySize() {
        ensureOpen();
        return PQC.nativeKemPublicKeySize(handle);
    }

    /**
     * Return the secret key size in bytes.
     */
    public int secretKeySize() {
        ensureOpen();
        return PQC.nativeKemSecretKeySize(handle);
    }

    /**
     * Return the ciphertext size in bytes.
     */
    public int ciphertextSize() {
        ensureOpen();
        return PQC.nativeKemCiphertextSize(handle);
    }

    /**
     * Return the shared secret size in bytes.
     */
    public int sharedSecretSize() {
        ensureOpen();
        return PQC.nativeKemSharedSecretSize(handle);
    }

    /**
     * Return the NIST security level (1-5).
     */
    public int securityLevel() {
        ensureOpen();
        return PQC.nativeKemSecurityLevel(handle);
    }

    /**
     * Generate a new keypair.
     *
     * @return the generated key pair
     * @throws PQCException on failure
     */
    public KeyPair keygen() throws PQCException {
        ensureOpen();
        byte[][] result = PQC.nativeKemKeygen(handle);
        if (result == null || result.length != 2) {
            throw new PQCException(PQCException.INTERNAL, "keygen returned unexpected result");
        }
        return new KeyPair(result[0], result[1]);
    }

    /**
     * Encapsulate a shared secret using the given public key.
     *
     * @param publicKey the recipient's public key
     * @return the encapsulation result (ciphertext + shared secret)
     * @throws PQCException on failure
     */
    public EncapsResult encaps(byte[] publicKey) throws PQCException {
        ensureOpen();
        if (publicKey == null || publicKey.length != publicKeySize()) {
            throw new PQCException(PQCException.INVALID_ARGUMENT, "invalid public key length");
        }
        byte[][] result = PQC.nativeKemEncaps(handle, publicKey);
        if (result == null || result.length != 2) {
            throw new PQCException(PQCException.INTERNAL, "encaps returned unexpected result");
        }
        return new EncapsResult(result[0], result[1]);
    }

    /**
     * Decapsulate a ciphertext to recover the shared secret.
     *
     * @param ciphertext the ciphertext from encapsulation
     * @param secretKey  the recipient's secret key
     * @return the shared secret
     * @throws PQCException on failure
     */
    public byte[] decaps(byte[] ciphertext, byte[] secretKey) throws PQCException {
        ensureOpen();
        if (ciphertext == null || ciphertext.length != ciphertextSize()) {
            throw new PQCException(PQCException.INVALID_ARGUMENT, "invalid ciphertext length");
        }
        if (secretKey == null || secretKey.length != secretKeySize()) {
            throw new PQCException(PQCException.INVALID_ARGUMENT, "invalid secret key length");
        }
        byte[] ss = PQC.nativeKemDecaps(handle, ciphertext, secretKey);
        if (ss == null) {
            throw new PQCException(PQCException.DECAPSULATION_FAILED, "decapsulation failed");
        }
        return ss;
    }

    /**
     * Release the native KEM context.
     */
    @Override
    public void close() {
        if (handle != 0) {
            PQC.nativeKemFree(handle);
            handle = 0;
        }
    }
}

File: bindings/kotlin/src/main/kotlin/com/dyber/pqc/KEM.kt
/*
 * libpqc-dyber - Post-Quantum Cryptography Library
 * Copyright (c) 2024-2026 Dyber, Inc.
 * SPDX-License-Identifier: Apache-2.0 OR MIT
 *
 * Kotlin KEM (Key Encapsulation Mechanism) wrapper.
 */

package com.dyber.pqc

/**
 * KEM keypair.
 */
data class KEMKeyPair(
    val publicKey: ByteArray,
    val secretKey: ByteArray
)

/**
 * KEM encapsulation result.
 */
data class KEMEncapsResult(
    val ciphertext: ByteArray,
    val sharedSecret: ByteArray
)

/**
 * Key Encapsulation Mechanism (KEM) operations.
 *
 * @param algorithm Algorithm name, e.g. "ML-KEM-768"
 */
class KEM(val algorithm: String) : AutoCloseable {

    private val handle: Long

    init {
        handle = NativeLib.pqc_kem_new(algorithm)
        require(handle != 0L) { "Unsupported KEM algorithm: $algorithm" }
    }

    /** Public key size in bytes. */
    val publicKeySize: Int
        get() = NativeLib.pqc_kem_public_key_size(handle)

    /** Secret key size in bytes. */
    val secretKeySize: Int
        get() = NativeLib.pqc_kem_secret_key_size(handle)

    /** Ciphertext size in bytes. */
    val ciphertextSize: Int
        get() = NativeLib.pqc_kem_ciphertext_size(handle)

    /** Shared secret size in bytes. */
    val sharedSecretSize: Int
        get() = NativeLib.pqc_kem_shared_secret_size(handle)

    /** NIST security level (1-5). */
    val securityLevel: Int
        get() = NativeLib.pqc_kem_security_level(handle)

    /**
     * Generate a new keypair.
     */
    fun keygen(): KEMKeyPair {
        val pk = ByteArray(publicKeySize)
        val sk = ByteArray(secretKeySize)
        val rc = NativeLib.pqc_kem_keygen(handle, pk, sk)
        check(rc == 0) { "KEM keygen failed: ${NativeLib.pqc_status_string(rc)}" }
        return KEMKeyPair(pk, sk)
    }

    /**
     * Encapsulate: generate shared secret and ciphertext from a public key.
     */
    fun encaps(publicKey: ByteArray): KEMEncapsResult {
        require(publicKey.size == publicKeySize) {
            "Public key must be $publicKeySize bytes, got ${publicKey.size}"
        }
        val ct = ByteArray(ciphertextSize)
        val ss = ByteArray(sharedSecretSize)
        val rc = NativeLib.pqc_kem_encaps(handle, ct, ss, publicKey)
        check(rc == 0) { "KEM encaps failed: ${NativeLib.pqc_status_string(rc)}" }
        return KEMEncapsResult(ct, ss)
    }

    /**
     * Decapsulate: recover shared secret from ciphertext using secret key.
     */
    fun decaps(ciphertext: ByteArray, secretKey: ByteArray): ByteArray {
        require(ciphertext.size == ciphertextSize) {
            "Ciphertext must be $ciphertextSize bytes, got ${ciphertext.size}"
        }
        require(secretKey.size == secretKeySize) {
            "Secret key must be $secretKeySize bytes, got ${secretKey.size}"
        }
        val ss = ByteArray(sharedSecretSize)
        val rc = NativeLib.pqc_kem_decaps(handle, ss, ciphertext, secretKey)
        check(rc == 0) { "KEM decaps failed: ${NativeLib.pqc_status_string(rc)}" }
        return ss
    }

    override fun close() {
        NativeLib.pqc_kem_free(handle)
    }

    companion object {
        /** Well-known algorithm names. */
        const val ML_KEM_512 = "ML-KEM-512"
        const val ML_KEM_768 = "ML-KEM-768"
        const val ML_KEM_1024 = "ML-KEM-1024"
        const val HQC_128 = "HQC-128"
        const val HQC_192 = "HQC-192"
        const val HQC_256 = "HQC-256"

        /** List all available KEM algorithms. */
        fun algorithms(): List<String> {
            val count = NativeLib.pqc_kem_algorithm_count()
            return (0 until count).map { NativeLib.pqc_kem_algorithm_name(it) }
        }
    }
}

/**
 * JNI native method declarations.
 * Delegates to the Java JNI binding in the java module.
 */
internal object NativeLib {
    init {
        System.loadLibrary("pqc_jni")
    }

    @JvmStatic external fun pqc_init(): Int
    @JvmStatic external fun pqc_cleanup()
    @JvmStatic external fun pqc_version(): String
    @JvmStatic external fun pqc_status_string(status: Int): String

    @JvmStatic external fun pqc_kem_algorithm_count(): Int
    @JvmStatic external fun pqc_kem_algorithm_name(index: Int): String
    @JvmStatic external fun pqc_kem_is_enabled(algorithm: String): Boolean

    @JvmStatic external fun pqc_kem_new(algorithm: String): Long
    @JvmStatic external fun pqc_kem_free(handle: Long)
    @JvmStatic external fun pqc_kem_public_key_size(handle: Long): Int
    @JvmStatic external fun pqc_kem_secret_key_size(handle: Long): Int
    @JvmStatic external fun pqc_kem_ciphertext_size(handle: Long): Int
    @JvmStatic external fun pqc_kem_shared_secret_size(handle: Long): Int
    @JvmStatic external fun pqc_kem_security_level(handle: Long): Int
    @JvmStatic external fun pqc_kem_keygen(handle: Long, pk: ByteArray, sk: ByteArray): Int
    @JvmStatic external fun pqc_kem_encaps(handle: Long, ct: ByteArray, ss: ByteArray, pk: ByteArray): Int
    @JvmStatic external fun pqc_kem_decaps(handle: Long, ss: ByteArray, ct: ByteArray, sk: ByteArray): Int

    @JvmStatic external fun pqc_sig_algorithm_count(): Int
    @JvmStatic external fun pqc_sig_algorithm_name(index: Int): String
    @JvmStatic external fun pqc_sig_is_enabled(algorithm: String): Boolean

    @JvmStatic external fun pqc_sig_new(algorithm: String): Long
    @JvmStatic external fun pqc_sig_free(handle: Long)
    @JvmStatic external fun pqc_sig_public_key_size(handle: Long): Int
    @JvmStatic external fun pqc_sig_secret_key_size(handle: Long): Int
    @JvmStatic external fun pqc_sig_max_signature_size(handle: Long): Int
    @JvmStatic external fun pqc_sig_security_level(handle: Long): Int
    @JvmStatic external fun pqc_sig_is_stateful(handle: Long): Boolean
    @JvmStatic external fun pqc_sig_keygen(handle: Long, pk: ByteArray, sk: ByteArray): Int
    @JvmStatic external fun pqc_sig_sign(handle: Long, sig: ByteArray, msg: ByteArray, sk: ByteArray): ByteArray
    @JvmStatic external fun pqc_sig_verify(handle: Long, msg: ByteArray, sig: ByteArray, pk: ByteArray): Boolean
}
